ion-schema/src/isl/version_based_isl/mod.rs

```diff
@@ -0,0 +1,480 @@
+//! Provides a way to construct ISL types and constraints programmatically.
+//!
+//! This module consists of three submodules that help constructing ISL types/constraint:
+//!
+//! * `constraints` module provides schema constraints [IslConstraint]
+//! which converts given ion content in the schema file into an ISL constraint(not-yet-resolved constraints).
+//!
+//! [IslConstraint]: IslConstraint
+//!
+//! ## Example usage of `version_based_isl` module to create an `IslSchema` using `IslType` for ISl 1.0:
+//!```
+//! use ion_rs::element::Element;
+//! use ion_schema::isl::version_based_isl::constraints::IslConstraint;
+//! use ion_schema::isl::version_based_isl::{IslSchema, IslSchemaHeader, IslTypeArgument, IslV1_0, UserReservedFields};
+//! use ion_schema::isl::version_based_isl::constraints::open_content::OpenContent;
+//! use crate::ion_schema::isl::version_based_isl::IslType;
+//! use crate::ion_schema::isl::version_based_isl::constraints::r#type::Type;
+//!
+//! // below code represents an ISL type:
+//! // type:: {
+//! //      name:my_type_name,
+//! //      type: int,
+//! //      all_of: [
+//! //          { type: bool }
+//! //      ],
+//! //      open_content: "This is an open content!"
+//! //  }
+//! let isl_type: IslType<IslV1_0> = IslType::named(
+//!     // represents the `name` of the defined type
+//!     "my_type_name".to_owned(),
```

Comment:
As a `TODO`, all our constructors should accept `impl Into<String>` and similar to avoid requiring boilerplate at the callsites.